### PR TITLE
Look for http status in header and convert to appropriate grpc status

### LINKED ISSUE #896: Implement HTTP→gRPC status code mapping
As defined in grpc/grpc#4955. Please wait until PR is in before merging the code changes, although the PR has received general agreement ("but it's not official until it's official").

This is important to return useful gRPC status codes when proxies generate errors.

### LINKED ISSUE #1190: Status 404 does not trigger an error on the client side
Version: 1.2.1
I deploy GRPC client and server with a http2-aware proxy in the middle. In some cases, my proxy returns status 404 to the client when the server is not available. GRPC client does not detect this as an error and returns an empty response with an empty error.
This seems like a bug, since GRPC mandates status 200 and status 404 should trigger a protocol error.

### PR DESCRIPTION
**Pull request by MakMukhi**
fixes #896
fixes #1190

Http to gRPC status code mapping : 
https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md

### CODE REVIEW

**Comment by hsaliak** [comment]
@MakMukhi what's remaining to commit this change in? Is it just waiting for a reviewer?

File: transport/http_util.go [hunk 1]
```diff
@@ -95,6 +95,12 @@ type decodeState struct {
 	err error // first error encountered decoding
 
 	encoding string
+	// Records if grpc-status field exists.
```
Comment (menghanl) [inline transport/http_util.go]: And reorder the variables.

File: transport/http2_client.go [hunk 2]
```diff
@@ -861,6 +862,24 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
 	for _, hf := range frame.Fields {
 		state.processHeaderField(hf)
 	}
+	// If grpc status doesn't exist.
+	if !state.statusExists {
+		// Check if http status exists.
+		if !state.hstatusExists {
```
Comment (menghanl) [inline transport/http2_client.go]: Maybe we should not try to read the content of state if stats.err != nil.
For example in the case where http-status is not a valid number and strconv.Atoi() returns an error.

File: transport/transport_test.go [hunk 3]
```diff
@@ -830,3 +831,84 @@ func TestIsReservedHeader(t *testing.T) {
...
+		defer func() {
+			done <- err
+		}()
+		conn, err := lis.Accept()
+		if err != nil {
+			t.Errorf("Error at server-side while accpeting: %v", err)
+			return
+		}
+		// Read the preface sent by client.
+		if _, err = io.ReadFull(conn, make([]byte, len(http2.ClientPreface))); err != nil {
```
Comment (menghanl) [inline transport/transport_test.go]: Is it easy to use net/http package instead?

File: transport/transport_test.go [hunk 4]
```diff
@@ -830,3 +831,84 @@ func TestIsReservedHeader(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("Error creating gRPC client: %v", err)
+	}
+	// Create a stream on this transport. This will also send headers to the server.
+	stream, err := tr.NewStream(context.Background(), &CallHdr{Method: "bogus/method", Flush: true})
+	if err != nil {
+		t.Fatalf("Error creating stream on client transport :%v", err)
+	}
+	// wait for the server to send header with http status or fail.
+	if err = <-done; err != nil {
```
Comment (menghanl) [inline transport/transport_test.go]: Is it OK to remove this?
stream.Read() should be a blocking call.

File: transport/transport_test.go [hunk 5]
```diff
@@ -830,3 +831,84 @@ func TestIsReservedHeader(t *testing.T) {
...
+		// Error already printed out once.
+		return
+	}
+	want := "code = " + fmt.Sprint(uint32(httpStatusConvTab[httpStatus]))
+	if _, err = stream.Read([]byte{}); err == nil {
+		t.Fatalf("Strean err is nil, want err with code: %s", want)
+	}
+	got := err.Error()
+	if !strings.Contains(got, want) {
```
Comment (menghanl) [inline transport/transport_test.go]: Check if the err is a StreamError, and compare err.Code?

File: transport/transport_test.go (hunk 3 above)
Comment (MakMukhi) [inline transport/transport_test.go]: We'll have to do the same thing inside the http server handle there, doesn't make it any easier to use that.

**Comment by MakMukhi** [comment]
Opening anotther PR.

**Pull request closed without merging**